### PR TITLE
Add moduli configuration guideline for OpenSSH

### PR DESCRIPTION
**Pull request by David-5-1**
Promote 3072-bit-long (and more) DH groups for modern configuration. (2048 for intermediate)
Remove /etc/ssh/moduli from secret files.

### CODE REVIEW

**Comment by floatingatoll** [comment]
I don't see any warning about breaking compatibility with Java 7 customers
in this patch. Is it still the case that preventing Logjam using > 512
moduli will break some Java users?

Shouldn we be promoting ECDH over DHE in this changeset as well?

Is 3072+ the appropriate real-world, client-compatible, recommendation for
use on mobile-accessed sites and with ECDH as well as DHE?

**Comment by David-5-1** [comment]
Regarding Java, I have no idea what SSH client are be widespread and especially in which version(s) (I found JSch and sshj whose latest release seem to support ECDH, so any change in moduli would not block their usage but cannot tell for actually used versions).

Paramiko may also require a look to check what algorithms it can use in widespread environments (it seems OK with the latest release).

I see no reason to promote ECDH over DH except for speed, which is coherent with TLS guidelines (where ECDH is explained to be disabled just for speed reasons). I have no real opinion about the gain of disabling it or not (disabling it would simplify all discussion on moduli file though :-) ).

Regarding the choice of 3072+ bits, it matches various governmental agencies advices for mid-term future (~2030) as summed up on https://www.keylength.com/en/5/ I can't say much more for I am not a cryptanalyst…

File: Security/Guidelines/OpenSSH.mediawiki [hunk 1]
```diff
@@ -74,6 +74,13 @@ PermitRootLogin No
 UsePrivilegeSeparation sandbox
 </source>
 
+File: <code>/etc/ssh/moduli</code>
+
+All Diffie-Hellman moduli in use should be at least 3072-bit-long (they are used for <code>diffie-hellman-group-exchange-sha256</code>). From the structure of <code>moduli</code> files, this means the fifth field of all lines in this file should be greater than or equal to 3071.
```
Comment (gdestuynder) [inline Security/Guidelines/OpenSSH.mediawiki]: would rephrase a bit for clarity & conciseness (same for subsequent paragraph):

> All Diffie-Hellman moduli in use should be at least 3072-bit-long (they are used for <code>diffie-hellman-group-exchange-sha256</code>) as per our [[Security/Guidelines/Key_Management]] recommendations.


(cut rest of paragraph)

File: Security/Guidelines/OpenSSH.mediawiki [hunk 2]
```diff
@@ -74,6 +74,13 @@ PermitRootLogin No
...
+
+To deactivate short modules in two commands: <code>awk '{if ($5 < 3071) {print ("#", $0)} else print $0}' /etc/ssh/moduli > /etc/ssh/moduli.tmp; mv /etc/ssh/moduli.tmp /etc/ssh/moduli</code>
```
Comment (gdestuynder) [inline Security/Guidelines/OpenSSH.mediawiki]: To replace the cut paragraph above, would add:

> See also  <code>man moduli</code>.

**Comment by gdestuynder** [comment]
Note: 3072bit is coherent with https://wiki.mozilla.org/Security/Guidelines/Key_Management

File: Security/Guidelines/OpenSSH.mediawiki [hunk 3]
```diff
@@ -472,3 +487,4 @@ However, it seems safe to say that the latency differences are not significant a
 * [https://security.googleblog.com/2014/04/speeding-up-and-strengthening-https.html CHACHA20 vs AES-GCM performance study]
 * [http://cvsweb.openbsd.org/cgi-bin/cvsweb/~checkout~/src/usr.bin/ssh/PROTOCOL.certkeys?rev=1.9&content-type=text/plain PROTOCOL.certkeys]
 * [https://wiki.gnupg.org/rfc4880bis rfc44880bis from GnuPG]
+* [https://weakdh.org/ Weak Diffie-Hellman and the Logjam Attack]
```
Comment (gdestuynder) [inline Security/Guidelines/OpenSSH.mediawiki]: Would also link to https://jbeekman.nl/blog/2015/05/ssh-logjam/ (perhaps on the same line even)

File: Security/Guidelines/OpenSSH.mediawiki (hunk 2 above)
Comment (gdestuynder) [inline Security/Guidelines/OpenSSH.mediawiki]: Clearer/more straight-forward command:

`awk '$5 >= 3072' /etc/ssh/moduli > /tmp/moduli.strong && \
mv /tmp/moduli.strong /etc/ssh/moduli`

File: Security/Guidelines/OpenSSH.mediawiki [hunk 4]
```diff
@@ -176,7 +189,7 @@ account    required     pam_nologin.so
 
 * NIST curves (<code>ecdh-sha2-nistp512,ecdh-sha2-nistp384,ecdh-sha2-nistp256</code>) are listed for compatibility, but the use of <code>curve25519</code> is [https://safecurves.cr.yp.to/ generally preferred].
 
-* SSH protocol 2 supports [https://en.wikipedia.org/wiki/Diffie%E2%80%93Hellman_key_exchange DH] and [https://en.wikipedia.org/wiki/Elliptic_curve_Diffie%E2%80%93Hellman ECDH] key-exchange as well as [https://en.wikipedia.org/wiki/Forward_secrecy forward secrecy].
+* SSH protocol 2 supports [https://en.wikipedia.org/wiki/Diffie%E2%80%93Hellman_key_exchange DH] and [https://en.wikipedia.org/wiki/Elliptic_curve_Diffie%E2%80%93Hellman ECDH] key-exchange as well as [https://en.wikipedia.org/wiki/Forward_secrecy forward secrecy]. When DH (not ECDH) is used, the moduli the server proposes should be at least 2048-bit-long and 3072-bit-long whenever possible.
```
Comment (gdestuynder) [inline Security/Guidelines/OpenSSH.mediawiki]: i would throw this change out as this duplicates https://wiki.mozilla.org/Security/Guidelines/Key_Management

File: Security/Guidelines/OpenSSH.mediawiki [hunk 5]
```diff
@@ -433,7 +446,9 @@ Key material identifies the cryptographic secrets that compose a key. All key ma
 This includes:
 * OpenSSH server keys (<code>/etc/ssh/ssh_host_*key</code>)
 * Client keys (<code>~/.ssh/id_{rsa,dsa,ecdsa,ed25519}</code> and <code>~/.ssh/identity</code>).
-* <code>/etc/ssh/moduli</code> also contains prime numbers and generators for use in the Diffie-Hellman key exchange and must be handled like key material.
+
+This do not include:
```
Comment (gdestuynder) [inline Security/Guidelines/OpenSSH.mediawiki]: would just remove the line instead

Comment (David-5-1) [inline Security/Guidelines/OpenSSH.mediawiki]: I have left it to make it clear this file had not been forgotten and to make it clear the previous version was incorrect. However, I will not oppose very hard its removal if some more people ask for it.

File: Security/Guidelines/OpenSSH.mediawiki (hunk 2 above)
Comment (David-5-1) [inline Security/Guidelines/OpenSSH.mediawiki]: Nice one, I did not know awk could be used like that :-)

Comment (David-5-1) [inline Security/Guidelines/OpenSSH.mediawiki]: For the record, 3071 is not a typo (I have no idea why but sizes seems to be indicated minus 1…)

File: Security/Guidelines/OpenSSH.mediawiki (hunk 4 above)
Comment (David-5-1) [inline Security/Guidelines/OpenSSH.mediawiki]: I could not find anything related to DH and ECDH but it should probably be added to that page rather than on several pages for maintainability. Linking anyway for now.

File: Security/Guidelines/OpenSSH.mediawiki (hunk 3 above)
Comment (David-5-1) [inline Security/Guidelines/OpenSSH.mediawiki]: Done on a new line to keep coherent with all links above.

**Comment by gdestuynder** [comment]
Thanks for your help and contribution
